custom_components/econnect_metronet/coordinator.py

```diff
@@ -0,0 +1,96 @@
+import logging
+from datetime import timedelta
+from typing import Any, Dict, Optional
+
+import async_timeout
+from elmo.api.client import ElmoClient
+from elmo.api.exceptions import InvalidToken
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    CONF_DOMAIN,
+    CONF_SCAN_INTERVAL,
+    CONF_SYSTEM_URL,
+    DOMAIN,
+    POLLING_TIMEOUT,
+    SCAN_INTERVAL_DEFAULT,
+)
+from .devices import AlarmDevice
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AlarmCoordinator(DataUpdateCoordinator):
+    def __init__(self, hass: HomeAssistant, config: ConfigEntry) -> None:
+        # Initialize all clients and devices
+        client = ElmoClient(config.data[CONF_SYSTEM_URL], config.data[CONF_DOMAIN])
+        scan_interval = config.options.get(CONF_SCAN_INTERVAL, SCAN_INTERVAL_DEFAULT)
+
+        # Attributes
+        self.device = AlarmDevice(client, config.options)
+
+        # Configure the coordinator
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=scan_interval),
+        )
+
+    async def _async_update_data(self) -> Optional[Dict[str, Any]]:
```

Comment:
Ideally, this must not use a polling strategy. In the end, this is a long-polling that behaves like a push system. We may think as a further refactoring (new PR) to remove HA polling and just use a loop that connects to the polling API, retrieves the data, and eventually publishes the data to entities using:
```python
coordinator.async_set_updated_data(data)
```

We can move forward with the current implementation, but eventually we should change that.

Reply:
See: https://developers.home-assistant.io/docs/integration_fetching_data/#pushing-api-endpoints